File: tests/test_list_iterator.py
from __future__ import annotations

import re
import shutil
import subprocess
import tempfile
import unittest
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]


class ListIteratorContractTests(unittest.TestCase):
    """Source-level guards for the container list iterator.

    extract_char() removes the element a FOR_EACH_CHARACTER loop is standing
    on whenever a mobile dies inside the loop, which happens in
    violence_update(), aggr_update(), char_update(), do_mindblast(),
    spell_earthquake() and about a dozen other places. list_remove() used to
    free the node immediately, leaving the iterator cursor -- which aliases
    node->next -- dangling. Because `next` is the last field of LIST_NODE,
    glibc leaves it readable in a freed chunk until that chunk is recycled, so
    the defect showed up only as rare unreproducible corruption rather than a
    reliable crash.
    """

    @classmethod
    def setUpClass(cls) -> None:
        cls.list_c = (ROOT / "src" / "list.c").read_text(encoding="utf-8")
        cls.list_h = (ROOT / "src" / "list.h").read_text(encoding="utf-8")
        cls.container = (ROOT / "src" / "container.c").read_text(encoding="utf-8")
        cls.comm = (ROOT / "src" / "comm.c").read_text(encoding="utf-8")

    def _list_remove_body(self) -> str:
        match = re.search(
            r"void list_remove\(.*?\n\{(?P<body>.*?)\n\}", self.list_c, re.DOTALL
        )
        self.assertIsNotNone(match, "could not locate list_remove()")
        return match.group("body")

    def test_list_remove_defers_the_free(self) -> None:
        body = self._list_remove_body()

        # The eager free is the bug. It must not come back.
        self.assertNotIn("free( node )", body)
        self.assertNotIn("free(node)", body)

        # Tombstone plus park for later reclamation.
        self.assertIn("node->data = NULL", body)
        self.assertIn("node->prev = list->pending", body)
        self.assertIn("list->pending = node", body)

    def test_pending_chain_must_not_reuse_the_next_pointer(self) -> None:
        """The parked chain has to hang off `prev`.

        A parked node's `next` is what lets a cursor sitting on it walk
        forward, so threading the free list through `next` would reintroduce
        the defect in a subtler form.
        """
        body = self._list_remove_body()
        self.assertNotIn("node->next = list->pending", body)
        self.assertIn("LIST_NODE *pending;", self.list_h)

    def test_iterator_skips_tombstoned_nodes(self) -> None:
        match = re.search(
            r"void \*list_iterator_next\(.*?\n\{(?P<body>.*?)\n\}",
            self.list_c,
            re.DOTALL,
        )
        self.assertIsNotNone(match)
        body = match.group("body")

        self.assertIn("while ( ( node = *iter->pnext ) != NULL )", body)
        self.assertIn("if ( node->data != NULL )", body)

    def test_flush_is_reachable_and_wired_into_the_game_loop(self) -> None:
        self.assertIn("void list_flush_pending( LIST *list )", self.list_c)
        self.assertIn("void    list_flush_pending( LIST *list );", self.list_h)

        # Both game-wide containers are reclaimed.
        self.assertIn("list_flush_pending( &character_list )", self.container)
        self.assertIn("list_flush_pending( &object_list )", self.container)

        # And the flush actually runs, once per pass of the main loop.
        self.assertIn("flush_container_lists();", self.comm)
        loop = self.comm[self.comm.index("void game_loop_unix"):]
        self.assertIn("flush_container_lists();", loop)

    def test_list_init_clears_the_pending_chain(self) -> None:
        match = re.search(
            r"void list_init\(.*?\n\{(?P<body>.*?)\n\}", self.list_c, re.DOTALL
        )
        self.assertIsNotNone(match)
        self.assertIn("list->pending = NULL", match.group("body"))

    @unittest.skipIf(shutil.which("gcc") is None, "gcc is not available")
    def test_behavioural_harness_passes_under_sanitizers(self) -> None:
        """Compile and run the C harness when a compiler is present.

        scripts/validate.sh runs this too; doing it here as well means a plain
        `python -m unittest` on a machine with gcc still catches a regression.
        """
        source = ROOT / "tests" / "test_list_iterator.c"
        self.assertTrue(source.is_file())

        with tempfile.TemporaryDirectory() as tmp:
            binary = Path(tmp) / "test_list_iterator"
            build = subprocess.run(
                [
                    "gcc",
                    "-g",
                    "-fsanitize=address,undefined",
                    f"-I{ROOT / 'src'}",
                    "-o",
                    str(binary),
                    str(source),
                    str(ROOT / "src" / "list.c"),
                ],
                capture_output=True,
                text=True,
            )
            if build.returncode != 0:
                self.skipTest(f"harness did not build here: {build.stderr[:400]}")

            run = subprocess.run(
                [str(binary)],
                capture_output=True,
                text=True,
                env={"ASAN_OPTIONS": "detect_leaks=1"},
            )
            self.assertEqual(
                run.returncode,
                0,
                f"iterator harness failed:\n{run.stdout}\n{run.stderr}",
            )
            self.assertIn("ALL PASS", run.stdout)


if __name__ == "__main__":
    unittest.main()

File: tests/test_list_iterator.c
/*
 * Regression test for the character/object list iterator.
 *
 * extract_char() removes the very element a FOR_EACH_CHARACTER loop is
 * standing on whenever a mobile dies inside the loop, which happens in
 * violence_update(), aggr_update(), do_mindblast(), spell_earthquake() and
 * about a dozen other places. list_remove() used to free the node
 * immediately, leaving the iterator cursor (which aliases node->next)
 * dangling. Because `next` is the last field of LIST_NODE, glibc's freed-chunk
 * metadata usually left it readable, so the bug surfaced only as rare
 * unreproducible corruption.
 *
 * Build and run this under AddressSanitizer; scripts/validate.sh does so.
 *   gcc -g -fsanitize=address,undefined -Isrc -o t tests/test_list_iterator.c src/list.c
 */
#include <stdio.h>
#include "list.h"

#define HAVE_FLUSH 1

static int failures = 0;

static void check(const char *what, int ok)
{
    printf("  %-58s %s\n", what, ok ? "PASS" : "FAIL");
    if (!ok) failures++;
}

/* Find the node holding `data` so we can remove it like remove_entry() does. */
static LIST_NODE *node_of(LIST *l, void *data)
{
    LIST_NODE *n;
    for (n = l->head; n != NULL; n = n->next)
        if (n->data == data) return n;
    return NULL;
}


/* Remove every remaining live node, then reclaim. Keeps the harness leak-free. */
static void teardown(LIST *l)
{
    while (l->head != NULL) list_remove(l, l->head);
#ifdef HAVE_FLUSH
    list_flush_pending(l);
#endif
}

int main(void)
{
    static char A='A', B='B', C='C', D='D';
    LIST l;
    LIST_ITERATOR it;
    void *seen[8];
    int n, ok;
    void *v;

    /* Case 1: remove the node the cursor is standing on (the UAF case). */
    list_init(&l);
    list_push_front(&l, &D); list_push_front(&l, &C);
    list_push_front(&l, &B); list_push_front(&l, &A);   /* A B C D */
    n = 0;
    list_iterator_start(&it, &l);
    while ((v = list_iterator_next(&it)) != NULL) {
        seen[n++] = v;
        if (v == &B) list_remove(&l, node_of(&l, &B));  /* remove CURRENT */
        if (n > 6) break;
    }
    ok = (n == 4 && seen[0]==&A && seen[1]==&B && seen[2]==&C && seen[3]==&D);
    check("remove current node: iteration completes correctly", ok);
    if (!ok) { int i; printf("    got %d:", n); for(i=0;i<n;i++) printf(" %c", *(char*)seen[i]); printf("\n"); }

    teardown(&l);

    /* Case 2: remove the node just ahead of the cursor -- must be skipped. */
    list_init(&l);
    list_push_front(&l, &D); list_push_front(&l, &C);
    list_push_front(&l, &B); list_push_front(&l, &A);
    n = 0;
    list_iterator_start(&it, &l);
    while ((v = list_iterator_next(&it)) != NULL) {
        seen[n++] = v;
        if (v == &B) list_remove(&l, node_of(&l, &C));  /* remove NEXT */
        if (n > 6) break;
    }
    ok = (n == 3 && seen[0]==&A && seen[1]==&B && seen[2]==&D);
    check("remove next node: it is skipped", ok);

    teardown(&l);

    /* Case 3: remove current AND the one ahead in the same step. */
    list_init(&l);
    list_push_front(&l, &D); list_push_front(&l, &C);
    list_push_front(&l, &B); list_push_front(&l, &A);
    n = 0;
    list_iterator_start(&it, &l);
    while ((v = list_iterator_next(&it)) != NULL) {
        seen[n++] = v;
        if (v == &B) { list_remove(&l, node_of(&l,&B)); list_remove(&l, node_of(&l,&C)); }
        if (n > 6) break;
    }
    ok = (n == 3 && seen[0]==&A && seen[1]==&B && seen[2]==&D);
    check("remove current + next: no removed element handed back", ok);
    if (!ok) { int i; printf("    got %d:", n); for(i=0;i<n;i++) printf(" %c", *(char*)seen[i]); printf("\n"); }

    teardown(&l);

    /* Case 4: removing every element while iterating (mass extinction). */
    list_init(&l);
    list_push_front(&l, &D); list_push_front(&l, &C);
    list_push_front(&l, &B); list_push_front(&l, &A);
    n = 0;
    list_iterator_start(&it, &l);
    while ((v = list_iterator_next(&it)) != NULL) {
        seen[n++] = v;
        list_remove(&l, node_of(&l, v));
        if (n > 6) break;
    }
    ok = (n == 4 && l.size == 0 && l.head == NULL);
    check("remove every node while iterating", ok);

    /* Case 5: flush reclaims memory and leaves the live list usable. */
#ifdef HAVE_FLUSH
    list_flush_pending(&l);
    check("flush leaves an empty, usable list", l.head == NULL && l.size == 0);
    list_push_front(&l, &A);
    list_iterator_start(&it, &l);
    check("list still works after a flush", list_iterator_next(&it) == &A);
    list_flush_pending(&l);
#endif

    teardown(&l);

    printf("\n%s (%d failure(s))\n", failures ? "FAILED" : "ALL PASS", failures);
    return failures != 0;
}
